Windows 11: introduce support for Voice Access (NVDA installed on Windows 11 Version 22H2 and later)

josephsl opened the pull request:
One of the possibilities after #18220 merge:

### Link to issue number:
Closes #16862 
Closes #17384 

### Summary of the issue:
NVDA does not support Windows 11 Voice Access features including text dictation announcement from everywhere.

### Description of user facing changes:
Introduces support for Windows 11 Voice Access, including text dictation, microphone toggle announcement, and mouse and touch interaction. Requires NVDA to be installed.

### Description of developer facing changes:
Added voiceaccess.py app module, based on #18220 work.

### Description of development approach:
The new Voice Access app module includes:

* isGoodUIAWindow: says "yes" for "Voice Access" window class name (enables mouse and touch interaction)
* Should process UIA notification event says "yes" as Voice Access notification elements do not have native window handle
* UIA notification event will announce display string from everywhere.

### Testing strategy:
Manual and requires NVDA to be installed:

1. Enable and run Voice Access on Windows 11.
2. When Voice Access screen appears, try toggling microphone by saying "microphone sleep". Observe that NVDA announces toggle state of microphone.
3. Move to a text field (say, Notepad) and dictate text. Observe that NVDA echoes back dictated text.

### Known issues with pull request:
As Voice Access requires admin privileges, this app module will not work on portable and temporary (installer) copies of NVDA. This limitation is documented in the user guide as an example of an app with admin privileges. Further, as dictated text will be announced from everywhere, users should be advised to wear headphones when dictating (for privacy) unless the situation does allow echoing of text to be acceptable.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

josephsl committed:
Windows 11 Voice Access: initial app module commit with UIA notificatoin processing accepted. Re #16862.

Windows 11 Version 22H2 (2022 Updat2/build 22621) introduces Voice Access as a replacement for Windows Speech Recognition. As part of its work, Voice Access raises UIA notification to announce messages including microphone toggle status and dictated text. However, these elements do not have native window handle, cuaisng NVDA to drop these events. Therefore, employing the just introduceg 'shouldProcessUIANotificationEvent' API, say 'yes' when Voice Access app module is asked to handle notification events. Doing so allows Voice Access status such as microphone toggle to be announced while the app is focused.

josephsl committed:
Windows 11 Voice Access: allow notifications to be announced from everywhere. Re #16862, #17384.

One of the functions of Voice Access is text dictation. In order to do so, a user must focus on a text field and start dictating. Interestingly, this means the user will not focus on Voice Access, causing NVDA to not announce dictated text because the base UIA object implementation stops notification event handling if focused somewhere other than the app the notification is coming from. Therefore, let Voice Access inform NVDA that it will announce display string (including dictated text) from the app module, thereby letting Voice Access notifications come through while using apps in scenarios such as text dictation.

josephsl committed:
Windows 11 Voice Access: allow mouse and touch interaction in Voice Access. Re #16862.

Allow moues and touch interaction to work in Voice Access interface (window class name: Voice Access).

josephsl committed:
What's new: document Windows 11 Voice Access support in 'new features' section. Re #16862, #17384

josephsl committed:
User guide: add Windows 11 Voice Access as an example of an application with admin privileges and thus requiring NVDA to be installed. Re #16862, #17384

seanbudd commented on source/appModules/voiceaccess.py:
```diff
@@ -0,0 +1,49 @@
+
+"""App module for Windows 11 Voice Access (Version 22H2 and later).
+Voice Access allows users to dictate text and perform voice commands, replacing Windows Speech Recognition.
+This app module cannot be used in portable version of NVDA."""
```
should there be an `isInstalledCopy` check somewhere?

seanbudd replied:
https://github.com/nvaccess/nvda/pull/18322#issuecomment-3002329517

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -475,7 +475,7 @@ This includes disabling usage of the [Add-on Store](#AddonsManager).
 
 * The inability to automatically start during and/or after log-on.
-* The inability to interact with applications running with administrative privileges, unless of course NVDA itself has been run also with these privileges (not recommended).
+* The inability to interact with applications running with administrative privileges including Command Prompt and PowerShell running in admin mode and Windows 11 Voice Access, unless of course NVDA itself has been run also with these privileges (not recommended).
```
can you turn this into a sublist?

seanbudd replied:
>https://github.com/nvaccess/nvda/pull/18322#issuecomment-3002331826

@josephsl - you would need to tag @Qchristensen to get this attention

josephsl commented:
Hi, I think this can be done either at import time (when the app module is imported) or as part of the constructor (__init__). AS noted in the user guide, portable NVDA release can access processes like Voice Access when run as an admin (discouraged), but I agree that we should do an install copy check in the ap module just to make sure. Thanks.

josephsl commented:
Hi,

On second thoughts, I think the comment on installed copy requirement can be removed as the app module won't work properly in portable NVDA anyway (this is also the case for Task manager and other apps that require admin privileges for full functionality). But the user guide example should be kept as Voice Access is indeed an app requiring admin rights for full functionality.

Thanks.

josephsl committed:
Merge branch 'master' into i16862w11VoiceAccess

josephsl committed:
Voice Access: remove code comment on installed copy requirement. Re #16862.

Voice Access app module will not work with portable NVDA due to lack of admin rights (UI Access) by default. While the source code comment on instlaled copy requirement is removed, the portable copy restriction will remain in the user guide (so that Voice Access can be used as an example of an app with portable copy restrictions).

josephsl committed:
Voice Access: announce notificaitons if there are thins to be announced. Re #16862.

Review form Sean Budd (NV Access): check to make sure there is something to say (display string is not None) before invoking ui.message.

josephsl committed:
User guide: separate admin rights apps examples to a separate list.

Recommended by Sean Budd (NV Access): split examples of apps requiring admin rights onto sublists for readability. That way future user guide edits can add more exmaples in additoin to Windows 11 Voice Access.

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -475,7 +475,9 @@ This includes disabling usage of the [Add-on Store](#AddonsManager).
 * The inability to automatically start during and/or after log-on.
-* The inability to interact with applications running with administrative privileges, unless of course NVDA itself has been run also with these privileges (not recommended).
+* The inability to interact with applications running with administrative privileges, unless of course NVDA itself has been run also with these privileges (not recommended). Examples include:
+  * Command Prompt and PowerShell running in admin mode
```
I think Task Manager is worth mentioning as we get complaints about that a lot

```suggestion
  * Command Prompt and PowerShell running in admin mode
  * Task Manager
```

Qchristensen reviewed (approved):
Looks good.

seanbudd committed:
Update user_docs/en/userGuide.md
